### PR TITLE
Support F# Type Providers resolution in C#

### PR DESCRIPTION
**Pull request by DedSec256**
image

JetBrains1
JetBrains1 (1)

### CODE REVIEW

**Review by worldbeater — APPROVED** [review]
This is so so cool, thanks for implementing this! JetBrains seem to be pioneers here 💯

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Util
+{
+  public static class SolutionExtension
+  {
+    public static bool TryGetProvidedType(this ISolution solution, IClrTypeName name,
+      out ProxyProvidedTypeWithContext providedType)
+    {
+      providedType = null;
+      var extensionTypingProvider = solution.GetComponent<IProxyExtensionTypingProvider>();
+      return extensionTypingProvider.TypeProvidersManager?.Context.ProvidedAbbreviations.TryGetValue(name.FullName,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs:15]: What is two different projects provide different but the same-abbreviated types? Shouldn't it take a psi module into account?

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+using JetBrains.Metadata.Reader.API;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Plugins.FSharp.Shim.TypeProviders;
+using JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Models;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Util
+{
+  public static class SolutionExtension
+  {
+    public static bool TryGetProvidedType(this ISolution solution, IClrTypeName name,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs:10]: Do we really want it as an extension method on a solution? It seems a bit too specific to type providers and not related to a solution in any way. Making it extension on a clr type name would seem a better fit, though maybe a non-extension method would be even better.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs (hunk 1 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/SolutionExtension.cs:15]: Yes, we have to take this into account, I'm working on it now.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/FSharpProvidedParameter.cs [hunk 3]
```diff
@@ -34,7 +35,10 @@ myInfo switch
         _ => ParameterKind.VALUE
       };
 
-    public bool IsParameterArray => false;
+    public bool IsParameterArray =>
+      myInfo is ProxyProvidedParameterInfoWithContext x &&
+      x.GetAttributeConstructorArgs(null, "System.ParamArrayAttribute") != null;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/FSharpProvidedParameter.cs:40]: Would it be easy to add tests for cases like this?

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/FSharpProvidedParameter.cs:40]: yes, there will be a test for that

**Review by auduchinok — APPROVED** [review]
LGTM, just a few things, @DedSec256 please take a look.

File: ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/Cache/ProvidedAbbreviationsCache.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+using JetBrains.ReSharper.TestRunner.Abstractions.Extensions;
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Cache
+{
+  public interface IProvidedAbbreviationsCache
+  {
+    void AddOrUpdate(ProxyProvidedTypeWithContext type);
+    bool TryGet(IPsiModule module, IClrTypeName clrName, out ProxyProvidedTypeWithContext providedType);
+    void MarkAsInvalidated(IPsiModule module, IClrTypeName clrName);
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/Cache/ProvidedAbbreviationsCache.cs:17]: This method does not invalidate the cache but adds an item to the queue for delayed invalidation

File: rider-fsharp/src/test/kotlin/typeProviders/TypeProvidersCSharpTest.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @Test
+    @TestEnvironment(
+        solution = "SwaggerProviderCSharp",
+        toolset = ToolsetVersion.TOOLSET_17_CORE,
+        coreVersion = CoreVersion.DOT_NET_6
+    )
+    fun changeStaticArg() {
+        withOpenedEditor(project, "SwaggerProviderLibrary/Literals.fs") {
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply rider-fsharp/src/test/kotlin/typeProviders/TypeProvidersCSharpTest.kt:49]: In this test, we dump errors in the C# file. Errors for F# files are checked in the usual tests for type providers.

File: ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/Utils/PrimitiveTypesBoxer.cs [hunk 6]
```diff
@@ -30,13 +35,14 @@ public static RdStaticArg BoxToServerStaticArg(object value)
...
+        // ReSharper disable once PossibleInvalidCastException
+        _ when value.GetType() is var type && type.IsEnum => BoxToServerStaticArg((int)value),
         _ => throw new ArgumentException($"Unexpected static arg with type {value.GetType().FullName}")
       };
     }
 
-    [ContractAnnotation("null => null")]
-    public static ServerRdStaticArg BoxToClientStaticArg(object value)
+    [ContractAnnotation("value:null => null")]
+    public static ServerRdStaticArg BoxToClientStaticArg(object value, bool safeMode = false)
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/Utils/PrimitiveTypesBoxer.cs:45]: Oh yes, it's not required. Forgot to remove, thanks (:

**Comment by auduchinok** [comment]
@DedSec256 Congrats on having this merged in! This was a lot of work to do. 🙂

**Pull request merged**